### PR TITLE
[Platform] Add MPS (Apple Metal) platform support for macOS

### PR DESCRIPTION
**Pull request by robtaylor**
## Summary

Add GPU-accelerated LLM inference on Apple Silicon Macs via the MPS (Metal Performance Shaders) backend. This addresses feature request #1441 (86 reactions).

**6 commits:**

1. **Core MPS platform** — `MpsPlatform`, MPS attention backend (pure PyTorch SDPA), worker, model runner, distributed init (gloo/HashStore), KV cache memory management, CI workflow
2. **E2E tests** — 4 tests (distilgpt2, dummy weights, 2-layer configs) + 38 attention unit tests
3. **Benchmarking script** — vLLM MPS vs llama.cpp comparison tool
4. **INT4 dequantization** — MPS branches in AWQ/GPTQ quantization paths using optional Metal kernels (11-13x speedup over PyTorch fallback)
5. **GGUF dequantization** — MPS branches for GGUF quantized models (Q4_0, Q8_0, Q4_K + more), graceful numpy fallback for unsupported types
6. **Installation docs** — "Apple MPS" tab in GPU installation guide with setup, usage, performance, troubleshooting

### Performance (Apple Silicon)

| Model | Quantization | Throughput |
|-------|-------------|------------|
| GGUF small model | Q8_0 | ~62 tok/s |
| GGUF small model | Q4_0 | ~45 tok/s |
| Qwen2.5-1.5B | INT4 AWQ | ~17 tok/s |
| Qwen2.5-1.5B | INT4 GPTQ | ~16 tok/s |

### Key design decisions

- **Pure PyTorch attention** — No C++ extensions needed for the attention backend; uses `torch.nn.functional.scaled_dot_product_attention` on MPS
- **Spawn multiprocessing** — `fork()` crashes on MPS; requires `VLLM_WORKER_MULTIPROC_METHOD=spawn`
- **Conservative memory default** — KV cache limited to 25% of system RAM to avoid Metal memory thrashing (50-100x slowdown above ~40%)
- **Optional Metal kernels** — INT4/GGUF Metal compute kernels are optional; quantized models work without them via PyTorch/numpy fallback (slower)
- **Unified memory model** — No discrete GPU memory; MPS shares system RAM

### Known limitations

- No PagedAttention on Metal (uses naive SDPA)
- No tensor parallelism (single GPU only)
- No continuous batching optimizations
- GGUF Q4_K_M models slow if model uses Q6_K layers (numpy fallback)
- Experimental status — best suited for single-user local inference

## Test plan

- [x] 38 attention unit tests pass on MPS
- [x] 4 E2E tests pass (distilgpt2, dummy weights)
- [x] INT4 AWQ/GPTQ inference validated (Qwen2.5-1.5B)
- [x] GGUF Q4_0/Q8_0/Q4_K inference validated
- [x] Memory thrashing regression verified (25% default avoids it)
- [ ] Verify docs render correctly with mkdocs

## Related

- #1441 — Feature request: Apple Silicon support (86 reactions)
- huggingface/kernels#308 — Metal kernel builder infrastructure
- Metal compute kernels: [rotary-embedding](https://github.com/robtaylor/rotary-embedding), [fused-rms-norm](https://github.com/robtaylor/fused-rms-norm), [dequant-int4](https://github.com/robtaylor/kernel-builder) (kernels-community)

### CODE REVIEW

File: docs/getting_started/installation/gpu.mps.inc.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+Currently, there are no pre-built MPS wheels. You must build from source.
+
+--8<-- [end:pre-built-wheels]
+--8<-- [start:build-wheel-from-source]
+
+Clone the MPS support branch and install:
+
+```bash
+git clone https://github.com/robtaylor/vllm.git
+cd vllm
```
[the remark this answers is not in the document]
Comment (robtaylor) [reply docs/getting_started/installation/gpu.mps.inc.md:35]: Fixed — now points to vllm-project/vllm main.

File: docs/getting_started/installation/gpu.mps.inc.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+```
+
+### Installing Metal quantization kernels (optional)
+
+For accelerated INT4 (AWQ/GPTQ) and GGUF inference, build and install the Metal dequantization kernels. These require [Nix](https://determinate.systems/nix-installer/) to build.
+
+```bash
+# INT4 dequantization (AWQ + GPTQ)
+cd kernels-community/dequant-int4
+nix build
```
[the remark this answers is not in the document]
Comment (robtaylor) [reply docs/getting_started/installation/gpu.mps.inc.md:57]: Fixed — changed to `torch*-metal-aarch64-darwin` wildcard.

File: docs/getting_started/installation/gpu.mps.inc.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+cp -r result/torch210-metal-aarch64-darwin/ \
+  $(python -c "import site; print(site.getsitepackages()[0])")/dequant_int4/
+
+# GGUF dequantization (Q4_0, Q8_0, Q4_K, and more)
+cd ../dequant-gguf
+nix build
```
[the remark this answers is not in the document]
Comment (robtaylor) [reply docs/getting_started/installation/gpu.mps.inc.md:63]: Fixed with wildcard, same as above.

File: vllm/v1/attention/backends/mps_attn.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+            attn_out = F.scaled_dot_product_attention(
+                q,
+                k,
+                v,
+                attn_mask=None,
+                dropout_p=0.0,
+                is_causal=(attn_metadata.causal and q_len > 1),
+                scale=self.scale,
+                enable_gqa=(self.num_heads != self.num_kv_heads),
```
[the remark this answers is not in the document]
Comment (robtaylor) [reply vllm/v1/attention/backends/mps_attn.py:326]: Acknowledged — the per-sequence loop is a known limitation documented in the PR description ("No PagedAttention on Metal"). Batching the SDPA call with padding + masking is the right next step, but we intentionally kept this simple for the initial implementation to get the platform plumbing reviewed first. A batched attention path (or a proper Metal PagedAttention kernel) would be a follow-up PR.
